Documentation for Indexed completion mode (#26958)
kubernetes/website: content/en/docs/tasks/job/indexed-parallel-processing-static.md

---
title: Indexed Job for Parallel Processing with Static Work Assignment
content_type: task
min-kubernetes-server-version: v1.21
weight: 30
---

{{< feature-state for_k8s_version="v1.21" state="alpha" >}}

<!-- overview -->

{{< note >}}
To be able to create Indexed Jobs, make sure to enable the `IndexedJob`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
on the [API server](docs/reference/command-line-tools-reference/kube-apiserver/)
and the [controller manager](/docs/reference/command-line-tools-reference/kube-controller-manager/).
{{< /note >}}

> In this example, we will run a Kubernetes Job with multiple parallel worker processes.
> Each worker Pod gets assigned a unit of work statically, which is determined by
> the pod index available in the {{< glossary_tooltip text="annotation" term_id="annotation" >}}
> `batch.alpha.kubernetes.io/job-completion-index`, also accessible to containers
> in the `JOB_COMPLETION_INDEX` environment variable.

Reviewer: Style / wording suggestion (_OK to merge as-is_):
```suggestion
In this example, you will run a Kubernetes Job that uses multiple parallel worker processes.
Each worker is a different container running in its own Pod. The Pods have an _index number_
that the control plane sets automatically, which allows each Pod to identify which part of the
overall task to work on.

There are two ways for the containerized task process to work out its own index number:

1. The pod index is available in the {{< glossary_tooltip text="annotation" term_id="annotation" >}}
   `batch.alpha.kubernetes.io/job-completion-index`; you can publish this information using the
   [downward API](/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#the-downward-api) mechanism.
2. When an `Indexed` Job creates Pods, the Pod's app containers have an environment variable
    set, `JOB_COMPLETION_INDEX`. You can look up this variable, which is a string representing
    the decimal value of the pod index for this working.
```

Here is an overview of the steps in this example:

1. **Create an image that can read the pod index**
2. **Start an Indexed Job**. The downward API allows you to pass the annotation
   as an environment variable or file to the container.

> 

Reviewer: This needs a prerequisites (“Before you begin”) section. See https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/#before-you-begin for an example (refer to `content/en/docs/tasks/inject-data-application/distribute-credentials-secure.md`).

For example:
```markdown
## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

This task uses the Docker command line tool to build a container image. 
You also need write access to a container image registry that can host the image
you build.

If you do not have Docker available on your computer, you should install it.
You can instead use a Docker-compatible tool to build container images from a 
`Dockerfile`, such as Podman. You would then need to change `docker` in the
examples to match the name of the tool you are using.
```

Author: Added, except the docker part for now.

<!-- steps -->

> ## Create an Image

Reviewer: nit:
```suggestion
## Create a container image
```

Lowercasing “Image” because people don't `kubectl create image` (and headings in documentation are sentence case).

To access the work item from the worker program, you have a few options:

> 1. Read the `JOB_COMPLETION_INDEX` environment variable. The Job controller

Reviewer: nit:
```suggestion
1. Read the `JOB_COMPLETION_INDEX` environment variable. The Job
   {{< glossary_tooltip text="controller" term_id="controller" >}}
```

   automatically links this variable to the annotation containing the completion
   index.
1. Read a file that contains the completion index.
1. Assuming that you can't modify the program, you can wrap it with a script
   that reads the index using any of the methods above and converts it into
   something that the program can use as input.
  

> In this example, we show option 3. The following program knows how to

Reviewer: (nit)
```suggestion
For this example, imagine that you chose option 3. The following program knows how to
```

accept a work item from a command line argument.

{{< codenew language="python" file="application/job/worker.py" >}}

Give the script execution permission:

```shell
chmod +x worker.py
```

> Now, build an image. Download the [Dockerfile](/examples/application/job/Dockerfile)

Reviewer: nit, because “image” has several meanings within IT but outside of Kubernetes
```suggestion
Now, build a container image. Download the [Dockerfile](/examples/application/job/Dockerfile)
```

Author: Done

and build the image with this command:

```shell
docker build -t job-worker .
```

For the [Docker Hub](https://hub.docker.com/), tag your app image with
your username and push to the Hub with the below commands. Replace
`<username>` with your Hub username.

```shell
docker tag job-worker <username>/job-worker
docker push <username>/job-worker
```

If you are using [Google Container Registry](https://cloud.google.com/tools/container-registry/),
tag your app image with your project ID, and push to GCR. Replace `<project>`
with your project ID.

```shell
docker tag job-wq-1 gcr.io/<project>/job-worker
gcloud docker -- push gcr.io/<project>/job-worker

> ```

Reviewer: This might not be in quite line with our content guide:
https://kubernetes.io/docs/contribute/style/content-guide/#third-party-content

Consider using [tabs](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#tabs) to document image push. That way if someone wants to add a tab for another kind of image registry then, within reason, they easily can. Order the tabs alphabetically.

## Define an Indexed Job

Here is a job definition. You'll need to edit the image to match name you used.

{{< codenew language="yaml" file="application/job/indexed-job.yaml" >}}

> In this example, we used the builtin `JOB_COMPLETION_INDEX` environment variable

Reviewer: style nit:
```suggestion
In this example, you use the builtin `JOB_COMPLETION_INDEX` environment variable
```

set by the job controller. Optionally, you can [define your own environment
variable via through the downward API](/docs/tasks/inject-data-application/environment-variable-expose-pod-information/):

{{< codenew language="yaml" file="application/job/indexed-job-env.yaml" >}}

If, instead, the worker program knows how to read a workitem from a file, you can
similarly [use the downward API to pass the annotation value as a volume file](/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#store-pod-fields).

If the worker program expects certain values (for example, filenames) instead of
obscure completion indexes, you can do the mapping in a simple script.

{{< codenew language="yaml" file="application/job/indexed-job-args.yaml" >}}

You can also load the list of values from a [ConfigMap as an environment variable

> or file](/docs/tasks/configure-pod-container/configure-pod-configmap/).

Reviewer: Maybe link to the init container concept here?

Author: You can do it without an init container, so it seems tangential to me.

## Running the Job

Now run the Job:

```shell
kubectl apply -f ./indexed-job.yaml
```

Wait a bit, then check on the job:

```shell
kubectl describe jobs/indexed-job
```

> 

Reviewer: We'd typically write:
```suggestion

The output is similar to:

```

Author: Done

Start Time:        Wed, 10 Mar 2021 19:01:31 +0000
Pods Statuses:     2 Running / 3 Succeeded / 2 Failed
Completed Indexes: 0-2
Pod Template:
  Labels:  controller-uid=cca0c0a9-13c4-4e91-a92e-b4f28e6240bc
           job-name=indexed-job
  Containers:
   worker:
    Image:      gcr.io/my-project/job-worker
    Port:       <none>
    Host Port:  <none>
    Command:
      bash
      -c
      items=(foo bar baz qux xyz)
      /worker.py ${items[$JOB_COMPLETION_INDEX]}

    Environment:  <none>
    Mounts:       <none>
  Volumes:        <none>
Events:
  Type    Reason            Age   From            Message
  ----    ------            ----  ----            -------
  Normal  SuccessfulCreate  14s   job-controller  Created pod: indexed-job-9vwg2
  Normal  SuccessfulCreate  13s   job-controller  Created pod: indexed-job-96n5k
  Normal  SuccessfulCreate  13s   job-controller  Created pod: indexed-job-49z8k
  Normal  SuccessfulCreate  1s    job-controller  Created pod: indexed-job-kqbsv
  Normal  SuccessfulCreate  1s    job-controller  Created pod: indexed-job-5m96r
```

In this example, we run the job with custom values for each index. You can
inspect their output:

```shell

> kubectl logs indexed-job-49z8k

Reviewer: ```suggestion
kubectl logs indexed-job-49z8k # Change this to match the name of a Pod in your cluster
```

you can publish the value of the annotation using the [downward API](/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#the-downward-api)
mechanism.
For convenience, the control plane automatically sets the downward API to
expose the index in the `JOB_COMPLETION_INDEX` environment variable.

Here is an overview of the steps in this example:

1. **Create an image that can read the pod index**. You might modify the worker
   program or add a script wrapper.
2. **Start an Indexed Job**. The downward API allows you to pass the annotation
   as an environment variable or file to the container.

## {{% heading "prerequisites" %}}

Be familiar with the basic,
non-parallel, use of [Job](/docs/concepts/workloads/controllers/job/).

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

To be able to create Indexed Jobs, make sure to enable the `IndexedJob`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
on the [API server](docs/reference/command-line-tools-reference/kube-apiserver/)
and the [controller manager](/docs/reference/command-line-tools-reference/kube-controller-manager/).

<!-- steps -->

> ## Create a container image

Reviewer: ```suggestion
## Choose an approach
```

Author: Done

To access the work item from the worker program, you have a few options:

1. Read the `JOB_COMPLETION_INDEX` environment variable. The Job
   {{< glossary_tooltip text="controller" term_id="controller" >}}
   automatically links this variable to the annotation containing the completion
   index.
1. Read a file that contains the completion index.
1. Assuming that you can't modify the program, you can wrap it with a script
   that reads the index using any of the methods above and converts it into
   something that the program can use as input.
  
For this example, imagine that you chose option 3 and you want to run the
[rev](https://man7.org/linux/man-pages/man1/rev.1.html) utility. This
program accepts a file as an argument and prints its content reversed.

```shell
rev data.txt
```

> This program is available in the [busybox container image](https://hub.docker.com/_/busybox):
> 
> {{< tabs name="busybox container image" >}}
> {{{< tab name="Docker Hub" >}}
> docker.io/library/busybox
> {{< /tab >}}
> {{< tab name="Google Container Registry" >}}
> mirror.gcr.io/library/busybox
> {{< /tab >}}}
> {{< /tabs >}}

Reviewer: As we're not pushing images, my earlier feedback doesn't apply and this can simplify to:
```suggestion
For this example, you'll use the `rev` tool from the [`busybox`](https://hub.docker.com/_/busybox) container image.
```

Author: Done

---
title: Indexed Job for Parallel Processing with Static Work Assignment
content_type: task
min-kubernetes-server-version: v1.21
weight: 30
---

{{< feature-state for_k8s_version="v1.21" state="alpha" >}}

<!-- overview -->

{{< note >}}
To be able to create Indexed Jobs, make sure to enable the `IndexedJob`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
on the [API server](docs/reference/command-line-tools-reference/kube-apiserver/)
and the [controller manager](/docs/reference/command-line-tools-reference/kube-controller-manager/).

> {{< /note >}}

Reviewer: I'd put this under prerequisites. I can see why it's here though.

Author: Prerequisites seems better. Moved.
